test/common/upstream/bounded_load_hlb_test.cc

```diff
@@ -0,0 +1,250 @@
+#include <memory>
+
+#include "envoy/config/cluster/v3/cluster.pb.h"
+
+#include "common/upstream/thread_aware_lb_impl.h"
+
+#include "test/common/upstream/utility.h"
+#include "test/mocks/upstream/mocks.h"
+
+namespace Envoy {
+namespace Upstream {
+namespace {
+
+class TestHashingLoadBalancer : public ThreadAwareLoadBalancerBase::HashingLoadBalancer {
+public:
+  TestHashingLoadBalancer(NormalizedHostWeightVectorConstPtr& ring) : ring_(std::move(ring)) {}
+  TestHashingLoadBalancer() : ring_(nullptr) {}
+  HostConstSharedPtr chooseHost(uint64_t hash, uint32_t /* attempt */) const override {
+    if (ring_ == nullptr) {
+      return nullptr;
+    }
+    return ring_->at(hash).first;
+  }
+  NormalizedHostWeightVectorConstPtr ring_;
+
+private:
+};
+
+using HostOverloadedPredicate = std::function<bool(HostConstSharedPtr host, double weight)>;
+class TestBoundedLoadHashingLoadBalancer
+    : public ThreadAwareLoadBalancerBase::BoundedLoadHashingLoadBalancer {
+public:
+  TestBoundedLoadHashingLoadBalancer(
+      ThreadAwareLoadBalancerBase::HashingLoadBalancerSharedPtr hlb_ptr,
+      NormalizedHostWeightVectorConstPtr& normalized_host_weights_ptr, uint32_t hash_balance_factor,
+      HostOverloadedPredicate is_host_overloaded)
+      : ThreadAwareLoadBalancerBase::BoundedLoadHashingLoadBalancer(
+            hlb_ptr, normalized_host_weights_ptr, hash_balance_factor),
+        is_host_overloaded_(is_host_overloaded) {}
+
+private:
+  HostOverloadedPredicate is_host_overloaded_;
+  bool isHostOverloaded(HostConstSharedPtr host, double weight) const override {
+    return is_host_overloaded_(host, weight);
+  }
+};
+
+class BoundedLoadHashingLoadBalancerTest : public testing::Test {
+public:
+  HostOverloadedPredicate getHostOverloadedPredicate(const std::vector<std::string> addresses) {
```

Comment:
i would pass this by const ref and remove the assignment on L51 and copy addresses into the lambda directly